cli/silence_add.go

```diff
@@ -0,0 +1,172 @@
+	'=~' then it will be assumed to be the value of the alertname pair.
+
+  amtool silence add 'alertname=~foo.*'
+
+	As well as direct equality, regex matching is also supported. The '=~' syntax
+	(similar to prometheus) is used to represent a regex match. Regex matching
+	can be used in combination with a direct match.
+
+  amtool silence add alertname=foo node={bar,baz}
+
+	This statement will create multiple silences. It will create a silence that
+	matches the alertname=foo label value pair and node=bar pair as well as a
+	silence that matches alertname=foo and node=baz.
+
+  amtool silence add alertname=foo{a,b} node={bar,baz}
+
+	Similar to the previous example this statement will create 4 silences to match
+	any combinartion of alertname=fooa or alertname=foob and node=bar or node=baz.
+	`,
+	RunE: add,
+}
+
+func init() {
+	user, _ := user.Current()
+	addCmd.Flags().StringP("author", "a", user.Username, "Username for CreatedBy field")
+	addCmd.Flags().StringP("expires", "e", "1h", "Duration of silence (100h)")
+	addCmd.Flags().String("expire-on", "", "Expire at a certain time (Overwrites expires) RFC3339 format 2006-01-02T15:04:05Z07:00")
+	addCmd.Flags().StringP("comment", "c", "", "A comment to help describe the silence")
+	viper.BindPFlag("author", addCmd.Flags().Lookup("author"))
+	viper.BindPFlag("expires", addCmd.Flags().Lookup("expires"))
+	viper.BindPFlag("comment", addCmd.Flags().Lookup("comment"))
+	viper.SetDefault("comment_required", false)
+	addFlags = addCmd.Flags()
+}
+
+func add(cmd *cobra.Command, args []string) error {
+	var err error
+
+	matchers, err := parseMatchers(args)
+	if err != nil {
+		return err
+	}
+
+	groups := parseMatcherGroups(matchers)
+
+	if len(groups) < 1 {
+		return errors.New("No matchers specified")
+	}
+
+	expire_on, err := addFlags.GetString("expire-on")
+	if err != nil {
+		return err
+	}
+
+	expires := viper.GetString("expires")
+	var endsAt time.Time
+
+	if expire_on != "" {
+		endsAt, err = time.Parse(time.RFC3339, expire_on)
+		if err != nil {
+			return err
+		}
+	} else {
+		duration, err := time.ParseDuration(expires)
+		if err != nil {
+			return err
+		}
+		endsAt = time.Now().UTC().Add(duration)
+	}
+
+	author := viper.GetString("author")
+	comment := viper.GetString("comment")
+	comment_required := viper.GetBool("comment_required")
+
+	if comment_required && comment == "" {
+		return errors.New("Comment required by config")
+	}
+
+	for _, matchers := range groups {
+		typeMatchers, err := TypeMatchers(matchers)
+		if err != nil {
+			return err
+		}
+		silence := types.Silence{
+			Matchers:  typeMatchers,
+			StartsAt:  time.Now().UTC(),
+			EndsAt:    endsAt,
+			CreatedBy: author,
+			Comment:   comment,
+		}
+
+		u, err := GetAlertmanagerURL()
+		if err != nil {
+			return err
+		}
+		u.Path = path.Join(u.Path, "/api/v1/silences")
+
+		buf := bytes.NewBuffer([]byte{})
+		enc := json.NewEncoder(buf)
+		err = enc.Encode(silence)
```

Comment:
`json.NewEncoder(buf).Encode(silence)`

The encode and decode stuff doesn't get re-used, so no need to assign it to a var.